lib/autoproj/ops/configuration.rb

```diff
@@ -1,9 +1,70 @@
+require "rgl/adjacency"
+require "rgl/dot"
+require "rgl/topsort"
+
 module Autoproj
     module Ops
         #--
         # NOTE: indentation is wrong to let git track the history properly
         #+++
 
+        # PackageSetHierachy be used to build the hierarchy of package set
+        # imports, as directed acyclic graph (DAG)
+        # so that they can be (topologically) sorted according to their
+        # dependencies
+        class PackageSetHierarchy
+            attr_reader :imports_of, :dag
+
+            def initialize(package_sets, root_pkg_set)
+                @dag = RGL::DirectedAdjacencyGraph.new
+
+                @imports_of = Hash.new
+                package_sets.each do |p|
+                    dependencies = PackageSet.resolve_imports(p)
```

Comment:
We don't need the recursive resolve_imports anymore, since the DAG does it naturally. Just iterate over all package sets and add edges for each direct dependency.

Reply:
It does only so if all the involved package sets are provided (or the tree depth is limited) and that does not seem to be guaranteed at that point?

Reply:
It is guaranteed at this point. I've created a commit to document the whole class, but I think I'll propose it as a new PR after this one gets merged.